lightning/src/routing/router.rs

```diff
@@ -349,63 +394,170 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+					res.push(new_entry.3);
+					cur_path_fees.push(new_entry.5);
+				}
+				res.last_mut().unwrap().fee_msat = final_value_msat;
+				res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
+
+				// Recompute fee considering the bottleneck.
+				// Also remember used amounts per channels to avoid relying on them in the next MPP iterations.
+				let mut recomputed_fee_msat = 0;
+				let mut total_proportional_fee_millionths = 0;
+				let mut i = 0;
+				for fees in cur_path_fees.clone() {
+					let current_hop_amount_msat = path_bottleneck_msat.unwrap().checked_add(recomputed_fee_msat);
+					let proportional_fee_millions = current_hop_amount_msat.unwrap().checked_mul(fees.proportional_millionths as u64);
+					let mut hop_to_update = res.get_mut(i).unwrap();
+					if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+							(fees.base_msat as u64).checked_add(part / 1000000) }) {
+
+						recomputed_fee_msat += new_fee;
+						total_proportional_fee_millionths += fees.proportional_millionths;
+						hop_to_update.fee_msat = new_fee;
+					}
+
+					*used_channels_with_amounts_msat.entry(hop_to_update.short_channel_id).or_insert(0) += path_bottleneck_msat.unwrap();
+					i += 1;
 				}
-				if res.last().unwrap().pubkey == *target {
-					break;
+
+				payment_paths.push((res, path_bottleneck_msat, total_proportional_fee_millionths, recomputed_fee_msat));
+				if let Some(chan_id_to_avoid) = smallest_channel_id {
+					channels_to_avoid.insert(chan_id_to_avoid);
 				}
+				break 'path_finding;
+			}
 
-				let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
-					Some(hop) => hop.3,
-					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination".to_owned(), action: ErrorAction::IgnoreError}),
-				};
-				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
-				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
-				res.push(new_entry);
+			match network.get_nodes().get(&pubkey) {
+				None => {},
+				Some(node) => {
+					add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
+				},
 			}
-			res.last_mut().unwrap().fee_msat = final_value_msat;
-			res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
-			let route = Route { paths: vec![res] };
-			log_trace!(logger, "Got route: {}", log_route!(route));
-			return Ok(route);
 		}
+		tries_left -= 1;
 
-		match network.get_nodes().get(&pubkey) {
-			None => {},
-			Some(node) => {
-				add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
-			},
+		if tries_left == 0 || found_amount_msat >= recommended_available_msat {
+			break 'route_finding;
+		}
+	}
+
+	if payment_paths.len() == 0 {
+		return Err(LightningError{err: "Failed to find a path to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	// Sort by total fees and take the best paths.
+	payment_paths.sort_by_key(|k| k.3);
```

Comment:
Honestly, I'm left very confused what's going on down here. Some comments would be super useful, as would breaking out the tuples into things with names.